src/java.base/share/classes/jdk/internal/classfile/impl/AttributeHolder.java

```diff
@@ -24,51 +24,70 @@
  */
 package jdk.internal.classfile.impl;
 
-import java.util.ArrayList;
-import java.util.List;
+import java.util.Arrays;
 
 import java.lang.classfile.Attribute;
 import java.lang.classfile.AttributeMapper;
 
 public class AttributeHolder {
-    private final List<Attribute<?>> attributes = new ArrayList<>();
+    private static Attribute<?>[] EMPTY_ATTRIBUTE_ARRAY = {};
+    private int attributesCount = 0;
+    private Attribute<?>[] attributes = EMPTY_ATTRIBUTE_ARRAY;
 
     public <A extends Attribute<A>> void withAttribute(Attribute<?> a) {
         if (a == null)
             return;
 
         @SuppressWarnings("unchecked")
         AttributeMapper<A> am = (AttributeMapper<A>) a.attributeMapper();
-        if (!am.allowMultiple() && isPresent(am)) {
-            remove(am);
+        int attributesCount = this.attributesCount;
+        var attributes = this.attributes;
+        if (!am.allowMultiple()) {
+            // remove if
+            for (int i = attributesCount - 1; i >= 0; i--) {
+                if (attributes[i].attributeMapper() == am) {
+                    attributesCount--;
+                    System.arraycopy(attributes, i + 1, attributes, i, attributesCount - i);
+                }
```

Comment:
```suggestion
                    break;
                }
```
We can just break when wew find the attribute as thia mapper already prevents the attribute from appearing more than once.```